### ISSUE TITLE
[Performance]: unconditional fp32 logits convert + aliasing copy costs ~12% of the greedy decode step (Gemma 2B, v5e, batch 512)

### ISSUE THREAD

**Issue by RubensZimbres** [vllm-project/tpu-inference#3127]
## Summary

In the non-spec-decode path, `_sample_from_logits` in `tpu_runner.py` converts the logits to float32 unconditionally before calling `sample()`. Measured consequences at batch 512 (Gemma 2B, TPU v5e-1, greedy decoding, offline batch), these are ranks 2 - 4 of the entire HLO op stats table for a steady-state generation trace on tpu-inference 0.24.0:

| Rank | Op | Shape | Avg time/step | Note |
|---|---|---|---|---|
| 2 | `copy.2` | f32[512,256000] | 1,562 µs | input-aliasing copy; identical layouts on both sides |
| 3 | `convert_element_type.1` | bf16→f32[512,256000] | 1,232 µs | standalone dispatch, own program id (`tpu_runner.py:1624` on 0.24.0) |
| 4 | `jit(sample)/reduce` (argmax) | f32[512,256000] | 696 µs | reads f32; ~2× the bytes it needs (`sampling.py:125` on 0.24.0) |

Total ≈ 3.1 ms of a ~26.5 ms decode step ≈ **12%**, on a workload class (greedy, high batch: structured output, extraction, agent traffic) where the sampling transforms are unused.

**Confirmed present on current main** as of ⟨sha⟩ (2026-07-08):

- unconditional `logits = logits.astype(jnp.float32)` in the non-spec branch of  `_sample_from_logits`: ⟨permalink 1⟩
- greedy branch of `sample()` returns its input (`ret_tokens = greedy_tokens; ret_logits = logits`), which forces the input-aliasing copy: ⟨permalink 2⟩

Note that `sample()` already performs its own `astype(jnp.float32)` after the argmax, so the runner-side convert is redundant in this path.

## Why fp32 is not needed for correctness here

bf16 to f32 widening is exact, so argmax over the widened tensor is mathematically identical to argmax over the original bf16, the greedy token cannot change. The f32 logits required by the logprobs consumers can be produced inside `sample()` (where the astype already exists), preserving the output dtype contract with no observable change downstream.

## Proposed change

In the greedy branch: run the argmax on the incoming bf16 logits; perform the single bf16 to f32 convert inside `sample()` to produce `ret_logits`, so the convert itself becomes the output materialization (eliminating the aliasing copy, since the output is no longer an aliased input); rebuild the warmup/precompiled graphs together with the dtype change (see the regression evidence in the discussion section below for why this last part is essential). Expected recovery: up to the ~12% above on greedy high-batch workloads, minus whatever the pipeline needs to retain.

### Report of performance regression

Not a regression. The cost is present and stable across the 0.23.0 release, the 0.24.0 release, and current main (⟨sha⟩). Filed as a proposal because it is a standing inefficiency in the greedy path, not a change in behavior between versions.

## Why this is filed as a proposal rather than submitted as a PR

I attempted the fix externally twice, on two releases, and both attempts regressed severely with the same signature:

- **0.23.0**: made the fp32 convert conditional on `do_sampling` at the output side to −14% throughput, batch TTFT ×5. Reverted; baseline reconfirmed.
- **0.24.0**: removed the runner-side input convert, letting `sample()`'s internal astype do the work to −26% throughput (19,320 → 14,258 tok/s), batch TTFT ×4.4 (275 → 1,209 ms). Reverted; baseline reconfirmed (19,217 tok/s, within session variance of the pre-patch 19,320).

**Diagnosis:** the precompiled graph menu built during the engine's warmup pass is keyed on the dtype at this program boundary. An external edit leaves the menu built for the old dtype, so runtime calls land off the precompiled paths on every step, which both the TTFT multiplication and the throughput loss detect. The fix therefore has to rekey the warmup/precompile pass together with the dtype change, i.e., it belongs at the pipeline level, inside the project. The measurements above are offered as evidence for that constraint, not only as failed attempts: any patch for this should be checked against batch TTFT, which detected the miscompilation both times before throughput was even read.

## Validation offer

I can validate a candidate fix on TPU v5e with same-session baseline discipline and the batch-TTFT regression check described above. The full measurement protocol (clean-run rules, treatment verification independent of the treatment flag, HLO shape-keyed trace analysis) is documented in DOI 10.5281/zenodo.21212010 and DOI 10.5281/zenodo.21227874, which measured this serving stack on the same hardware.

Related: #1424 (sampling-path performance).

### Your current environment (if you think it is necessary)

Name: tpu_inference
Version: 0.24.0
Name: vllm-tpu
Version: 0.24.0
Name: jax
Version: 0.10.2
Name: libtpu
Version: 0.0.42.1
Name: torch
Version: 2.10.0
Python 3.12.13

**Comment by jerviscz** [comment]
## Independent reproduction on TPU v5e-1, tpu-inference 0.26.0

I independently reproduced all three pathologies on a separate v5litepod-1, two releases after the original report, using the repo's own profiling harness:

```
python3 examples/tpu_profiling.py --model google/gemma-2b --input-len 32 --output-len 128 --batch-size 512 --max-num-seqs 512 --max-model-len 512 --max-num-batched-tokens 1024 --gpu-memory-utilization 0.75
```

(greedy via the script's built-in `temperature=0.0` + `ignore_eos`; the non-default memory flags work around two startup OOMs that v5e-1 hits with this 256K-vocab model — one already resolved on current main by #3274's `MAX_PRECOMPILE_PROMPT_TOKENS` guard. Neither affects the decode-path shapes measured here.)

**Environment:** tpu-inference 0.26.0 / vllm-tpu 0.26.0, jax 0.10.2, jaxlib 0.10.2, libtpu 0.0.43, Python 3.12.13, v5litepod-1 (16 GB HBM).

### Per-op numbers (mean over 128 occurrences, XProf HLO Op Stats)

| Op | Shape | This repro, 0.26.0 (mean) | Original report, 0.24.0 |
|---|---|---|---|
| `copy.2` | f32[512,256000] | 1,561.98 µs | 1,562 µs |
| `convert_element_type.1` | bf16→f32 [512,256000] | 1,236.7 µs | 1,232 µs |
| argmax (`iota_reduce_fusion`) | reads f32[512,256000] | 696.2 µs | 696 µs |

The rank claim also reproduces: these are **ranks 2–4 of the entire HLO op stats table** sorted by average time, behind only the lm_head matmul itself (rank 1, the `jit(compute_logits_func)` fusion, 2,796.7 µs — the op that legitimately produces the logits).

### Corroborating op-level detail (XProf op metadata)

- The convert runs as its own standalone program (`tf_op: jit(convert_element_type)`), `bytes_accessed` = 786,432,000 (262 MB bf16 read + 524 MB f32 write), achieving ~636 GB/s = 78% of v5e HBM peak. Source attribution points at the unconditional `astype` in `_sample_from_logits` (L1992 in the 0.26.0 wheel; https://github.com/vllm-project/tpu-inference/blob/a5596b27f/tpu_inference/runner/tpu_runner.py#L2005 on current main).
- `copy.2` has character-identical operand and result layouts (`f32[512,256000]{1,0:T(8,128)}` on both sides), `model_flops: 0`, `bytes_accessed` = 1,048,576,000, and its operand is the program parameter `%logits.1` — a pure input-aliasing materialization copy, exactly as described.
- The argmax reads the widened f32 logits (`bytes_accessed` = 524,291,072 per step vs the ~262 MB it would read in bf16), source https://github.com/vllm-project/tpu-inference/blob/a5596b27f/tpu_inference/layers/jax/sample/sampling.py#L125, achieving 753 GB/s = 92% of peak.

### Cost scales linearly with batch (free experiment courtesy of the warm-up ramp)

The batch ramp-up executed the smaller `num_reqs` padding buckets a few times each, so the op stats table contains a natural batch-size sweep of the same two ops:

| padded batch | `copy.2` mean | argmax mean |
|---|---|---|
| 512 | 1,562 µs | 696 µs |
| 256 | 781 µs | 350 µs |
| 128 | 396 µs | 177 µs |

Each halving of batch halves both costs — consistent with pure bandwidth-bound behavior, i.e. the overhead grows as batch × vocab. (I plan to confirm the vocab axis on a second large-vocab model next.)

### Step time and recoverable waste

Decode step time is ~27 ms derived from untraced full-pass wall time, ~36 ms while tracing (host-side tracer overhead inflates inter-program gaps; device op durations are unaffected). Counting the copy and the convert in full plus the argmax's f32 premium (696 − ~350 µs), recoverable waste ≈ 3.1 ms ≈ **11–12% of the greedy decode step**, matching the original report across a two-release gap.

One additional cross-path note: this run exercised the torchax fallback path (gemma-2b has no native JAX definition; the rank-1 fusion's source attribution points into `torchax/ops/jtorch.py`), and all three pathologies appear unchanged — consistent with the issue's claim that this is runner-level and model-implementation-agnostic.

### Next

I plan to work on the pipeline-level fix along the lines proposed here: move the convert inside `sample()` so the widening becomes the output materialization, conditionalize the post-sample convert on `logprobs`, and rebuild the precompiled sampling menu (`_precompile_sampling` currently bakes an f32 dummy: https://github.com/vllm-project/tpu-inference/blob/a5596b27f/tpu_inference/runner/compilation_manager.py#L966) with the model's actual logits dtype so the warmup keying stays consistent — checking batch TTFT first per the regression protocol above. I'll gladly take the validation offer once a PR is up.
